core/primitives/src/shard_layout.rs

```diff
@@ -83,17 +96,195 @@ pub struct ShardLayoutV1 {
+            if account_id < boundary_account {
+                break;
+            }
+            shard_id += 1;
+        }
+        shard_id
+    }
+}
+
+/// A boundary of an account range. Can be the start, the end or an account id
+/// in between two shards.
+#[derive(serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq, Ord)]
+pub enum AccountBoundary {
+    Start,
+    End,
+    Middle(AccountId),
+}
+
+impl fmt::Debug for AccountBoundary {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            AccountBoundary::Start => write!(f, "start"),
+            AccountBoundary::End => write!(f, "end"),
+            AccountBoundary::Middle(account_id) => write!(f, "{}", account_id.as_str()),
+        }
+    }
+}
+
+impl PartialOrd for AccountBoundary {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        match (self, other) {
+            (AccountBoundary::Start, AccountBoundary::Start) => Some(Ordering::Equal),
+            (AccountBoundary::End, AccountBoundary::End) => Some(Ordering::Equal),
+
+            (AccountBoundary::Start, AccountBoundary::End) => Some(Ordering::Less),
+            (AccountBoundary::End, AccountBoundary::Start) => Some(Ordering::Greater),
+
+            (AccountBoundary::Start, AccountBoundary::Middle(_)) => Some(Ordering::Less),
+            (AccountBoundary::Middle(_), AccountBoundary::Start) => Some(Ordering::Greater),
+
+            (AccountBoundary::End, AccountBoundary::Middle(_)) => Some(Ordering::Greater),
+            (AccountBoundary::Middle(_), AccountBoundary::End) => Some(Ordering::Less),
+
+            (AccountBoundary::Middle(a), AccountBoundary::Middle(b)) => a.partial_cmp(b),
+        }
+    }
+}
+
+/// The account range of a single shard.
+#[derive(serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq, Ord)]
+pub struct AccountRange {
+    start: AccountBoundary,
+    end: AccountBoundary,
+}
+
+impl PartialOrd for AccountRange {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        match self.start.partial_cmp(&other.start) {
+            Some(core::cmp::Ordering::Equal) => {}
+            ord => return ord,
+        }
+        self.end.partial_cmp(&other.end)
+    }
+}
+
+impl fmt::Debug for AccountRange {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let range = format!("{:?}..{:?}", self.start, self.end);
+        f.debug_struct("AccountRange").field("range", &range).finish()
+    }
+}
+
+impl AccountRange {
+    pub fn new_single_shard() -> Self {
+        Self { start: AccountBoundary::Start, end: AccountBoundary::End }
+    }
+
+    pub fn new_start(end: &str) -> Self {
+        Self { start: AccountBoundary::Start, end: AccountBoundary::Middle(end.parse().unwrap()) }
+    }
+
+    pub fn new_end(start: &str) -> Self {
+        Self { start: AccountBoundary::Middle(start.parse().unwrap()), end: AccountBoundary::End }
+    }
+
+    pub fn new_mid(start: &str, end: &str) -> Self {
```

Comment:
why is `&str` preferred to `&AccountId`?

Reply:
Technically it should be AccountId but it's just so much more conveninent to
pass &str because then I just need to do .parse().unwrap() in one place. I'm
happy to do it proper way and add some convenience functions just for the tests,
please let me now if you think it's worth it.

Reply:
I'm inclined to think it's worth it, just because all these methods are public, but I don't have a strong preference.